keps/sig-node/1977-container-notifier/README.md

```diff
@@ -0,0 +1,877 @@
+- Adding an optional field `Notifiers` which is a list of `ContainerNotifier`s into `Container`.
+- Adding an inlined type `ContainerNotifierHandler` which defines a command.
+- Adding a core API type `PodNotification` which defines request type to trigger execution of `ContainerNotifier`(s) in a `Pod`.
+- Introduce a new gate `ContainerNotifier` to toggle this feature.
+
+A SINGLE *trusted* controller(Pod Notification Controller) will be implemented to watch `PodNotification` resources, execute the command and update their statuses accordingly.
+
+In phase 2, propose to:
+- Add a core API `Notification` type and a controller which processes `Notification` resources.
+- Add an inline `Pod` definition for signals and allows the API object to send a request to trigger delivering of those signals.
+- Move logic in the Pod Notification controller into Kubelet. Kubelet watches `PodNotification` objects, runs the command and updates statuses of `PodNotification` objects accordingly.
+
+In phase 3, a `Probe` may be added if needed as an inline `Pod` definition to verify the signal is delivered or whether the command is run and results in the desired outcome.
+
+Details of proposed API changes can be found in *API Changes* section  bellow.
+
+### API Changes
+
+#### Phase 1 API Changes
+
+##### Inline Definition - ContainerNotifier
+
+Add []ContainerNotifier to the Container struct:
+
+```
+type Container struct {
+    ......
+    // +optional
+    Lifecycle *Lifecycle `json:"lifecycle,omitempty" protobuf:"bytes,12,opt,name=lifecycle"`
+
+    // Notifiers can be triggered by PodNotification resources.
+    // Each notifier must have a unique name within the Container.
+    // Upon creation of a PodNotification, Container(s) in the specified
+    // Pod with a matching notifier name will be executed.
+    // +optional
+    Notifiers []ContainerNotifier
+    ......
+}
+
+type ContainerNotifier struct {
+    // Name of the ContainerNotifier specified as a DNS_LABEL. Each
+    // ContainerNotifier within a Container must have a unique name.
+    // Names with the prefix `k8s.io/` are reserved for Kubernetes-defined
+    // "well-known" notifiers.
+    // Names with other prefixes such as `example.com/` are custom names.
+    // Immutable.
+    Name string
+
+    // The action to take when a matching PodNotification is observed.
+    // Exactly-once execution of the action is NOT guaranteed, so it's user's
+    // responsibility to make sure the handler is idempotent to get consistent
+    // behavior.
+    // For example, when a quiesce command has been ran on a database, the
+    // database will stay quiesced, and any consecutive execution of the
+    // same quiesce command should not have additional impact.
+    Handler *ContainerNotifierHandler
+
+    // Number of seconds after which the notifier will time out.
+    // Default to 1 second. Minimum value is 1.
+    // +optional
+    TimeoutSeconds int32
```

Comment:
Should there be a Handler that is run on timeout?

Reply:
How to handle timeout/errors is indeed an area that the current API proposal intentionally tries to simply and transfer some of the complexity to users. Could you give some concrete examples? One thing that bothers me for introducing  error/timeout handlers is that another "timeout" might be needed for the error/timeout handler itself. It would also complicate the API/implementation quite a bit. WDYT?

Reply:
I agree that it will complicate things, what I was worried about is if I quiesce an application but  it times out, to bring the application back up to running, we would want to run an unquiesce operation? 

I guess this could be solved by the notifier having a graceful termination with a shutdown signal being sent on timeout? Wondering if it should be apart of the enhancement?

Reply:
the main issue with introducing a graceful termination in the API is that it does not fundamentally solve the problem. It's possible the graceful termination mechanism itself is not graceful enough :). Meaning, it might end up as a failure. It seems to be reasonable to me that Applications/controllers understands the semantic of returned errors of a notification, and handle it appropriately (i.e., keep on retrying an un-quiesce hook until it succeeds).